helix-core/src/main/java/org/apache/helix/task/TaskUtil.java

```diff
@@ -220,6 +221,93 @@ protected static void setWorkflowContext(HelixManager manager, String workflowRe
+   *
+   * @param manager             a connection to Helix
+   * @param workflowJobResource the name of workflow
+   * @param key                 the key of key-value pair
+   *
+   * @return null if there is no such pair, otherwise return a String
+   */
+  protected static String getWorkflowJobUserContent(HelixManager manager,
+      String workflowJobResource, String key) {
+    ZNRecord r = manager.getHelixPropertyStore().get(Joiner.on("/")
+            .join(TaskConstants.REBALANCER_CONTEXT_ROOT, workflowJobResource, USER_CONTENT_NODE), null,
+        AccessOption.PERSISTENT);
+    return r != null ? r.getSimpleField(key) : null;
+  }
+
+  /**
+   * Add an user defined key-value pair data to workflow or job level
+   *
+   * @param manager             a connection to Helix
+   * @param workflowJobResource the name of workflow or job
+   * @param key                 the key of key-value pair
+   * @param value               the value of key-value pair
+   */
+  protected static void addWorkflowJobUserContent(HelixManager manager, String workflowJobResource,
+      String key, String value) {
+    String path = Joiner.on("/")
+        .join(TaskConstants.REBALANCER_CONTEXT_ROOT, workflowJobResource, USER_CONTENT_NODE);
+    ZNRecord r = manager.getHelixPropertyStore().get(path, null, AccessOption.PERSISTENT);
+    r.setSimpleField(key, value);
+    manager.getHelixPropertyStore().set(path, r, AccessOption.PERSISTENT);
+  }
+
+  /**
+   * Get user defined task level key-value pair data
+   *
+   * @param manager      a connection to Helix
+   * @param jobResource  the name of job
+   * @param taskResource the name of the task
+   * @param key          the key of key-value pair
+   *
+   * @return null if there is no such pair, otherwise return a String
+   */
+  protected static String getTaskUserContent(HelixManager manager, String jobResource,
+      String taskResource, String key) {
+    ZNRecord r = manager.getHelixPropertyStore().get(
+        Joiner.on("/").join(TaskConstants.REBALANCER_CONTEXT_ROOT, jobResource, USER_CONTENT_NODE),
+        null, AccessOption.PERSISTENT);
+    return r != null ? (r.getMapField(taskResource) != null
+        ? r.getMapField(taskResource).get(key)
+        : null) : null;
+  }
+
+  /**
+   * Add an user defined key-value pair data to task level
+   *
+   * @param manager       a connection to Helix
+   * @param jobResource   the name of job
+   * @param taskResource  the name of task
+   * @param key           the key of key-value pair
+   * @param value         the value of key-value pair
+   */
+  protected static void addTaskUserContent(HelixManager manager, String jobResource,
+      String taskResource, String key, String value) {
+    String path =
+        Joiner.on("/").join(TaskConstants.REBALANCER_CONTEXT_ROOT, jobResource, USER_CONTENT_NODE);
+    ZNRecord r = manager.getHelixPropertyStore().get(path, null, AccessOption.PERSISTENT);
```

Comment:
Is it possible that r == null?  Since you checked r == null in getTaskUserContent